dbos-rules.ts

```diff
@@ -531,18 +542,62 @@ const isSqlInjection: ErrorChecker = (node, fnDecl, isLocal) => {
       return checkCallForInjection(maybeArgs[0], fnDecl, isLocal);
     }
   }
-}
+};
+
+////////// This code is for detecting useless transactions
+
+// Note: this may result in false negatives for nested closures that capture the transaction context.
+const transactionDoesntUseTheDatabase: ErrorChecker = (node, fnDecl, _isLocal) => {
+  if (node !== fnDecl) return; // Only analyze the whole function
+
+  const params = fnDecl.getParameters();
+  if (params.length === 0) return; // In this case, not a valid transaction
+
+  const transactionContextSymbol = getSymbol(params[0]); // The first param should be the transaction context
+  if (transactionContextSymbol === Nothing) return; // No symbol for the first param -> should not analyze
+
+  //////////
+
+  let foundDatabaseUsage = false;
+
+  fnDecl.getBody()!.forEachDescendant((descendant, traversalControl) => {
+    if (transactionContextSymbol === getSymbol(descendant)) {
```

Comment:
Can you explain a bit more on how you detect database usage?

Does this only check the usage of `ctxt`? To detect whether a function is using the database, we want to make sure it uses `ctxt.client`

Reply:
Yes, this only checks for usage of  the passed-in transaction context. What the function does is that it checks if that symbol is present anywhere in the function body.

There's only one field behind 'ctxt', which is the client, so I'm not sure it would be really necessary to check for client field usage as well. And also, if you checked for just ctxt.client, then there would be a bunch of false positives for when transactions call helper functions that take just a context as a parameter.

Reply:
Does that make sense?

Reply:
Also my in-flight WiFi is limited so there might be slight communication gaps - sending this from an iPad; just 20 minutes WiFi per device so

Reply:
There are more fields behind ctxt: it contains all the properties from the base context (https://docs.dbos.dev/api-reference/contexts#properties).

However, I do agree that we don't want too many false positives.

Reply:
Got it that makes sense - didn't see that before
I'll change it to check for a `ctxt.client` or `client`, and see how many false positives arise over the demo apps

Reply:
Also, I have a question (unrelated to that):

So the definition for `PrismaClient` in `dbos-transact` is here:

https://github.com/dbos-inc/dbos-transact/blob/396d78eeeb51acf133fce10eeedf637e7989b916/src/user_database.ts#L178

But in the bank demo app backend, there's a different definition of `PrismaClient` which is used:

https://github.com/dbos-inc/dbos-demo-apps/blob/2f5cc06d43922667865db5bc3ff7e39b322f0d4b/bank/bank-backend/src/workflows/accountinfo.workflows.ts#L5

and

https://github.com/dbos-inc/dbos-demo-apps/blob/2f5cc06d43922667865db5bc3ff7e39b322f0d4b/bank/bank-backend/src/workflows/txnhistory.workflows.ts#L22

Should I use the official one from Prisma, or the interface from `dbos-transact`? Just wondering since my current mocked interface for it in the tests is using the official one.

Reply:
The tricky thing with PrismaClient is that it's auto-generated: https://www.prisma.io/docs/orm/prisma-client/setup-and-configuration/introduction
Unlike other ORMs which are installed directly, the PrismaClient is unique per application. Whenever you update your Prisma schema, you have to re-generate the client. Then DBOS will need to dynamically load the generated client ([here](https://github.com/dbos-inc/dbos-transact/blob/main/src/dbos-executor.ts#L230)), which is different per app.

Therefore, we need to provide a "mock" interface to capture common functions such as `$transaction` and `$queryRawUnsafe` for us to use Prisma internally.